docs/src/proposals/blockstore-rocksdb-compaction.md

```diff
@@ -0,0 +1,234 @@
+Although the deleted entries will not be visible from the read side right
+after the deletion is issued, the deleted entries (including the original
+data entries and its deletion entries) will still occupy disk storage.
+
+TBD: explain how write-key order makes this worse
+
+### High Write I/O from Write Amplification
+In addition to write stalls, it is also observed that compactions cause
+unwanted high write I/O.  With the current design where level compaction
+is configured for BlockStore, it has ~30x write amplification (10x write amp
+per level and assuming three levels in average).
+
+## Current Design
+In the current BlockStore, The default level compaction + slot-ID based
+compaction filter + periodic compaction are used, which has ~30x write
+amplification (10x write amp per level and assuming three levels in
+average + the write amp caused by the periodic compaction.)
+
+## Proposed Design -- FIFO Compaction with TTL + Rate Limiting
+As all the above issues are compaction related, it can be solved with
+a proper compaction style and deletion settings.  Fortunately, Solana's
+BlockStore has an unique workload where write-keys are mostly monotonically
+increasing over time.  This allows data to be persisted in sorted order
+naturally without compaction
+
+In the proposed design, we will laverage such unique property to aggressively
+config RocksDB to run as few compactions as possible while offering low read
+amplification with no write stalls.
+
+### FIFO Compaction
+FIFO Compaction is one of the three compaction styles offered by RocksDB which
+actually does not run any compaction in background.  Instead, it simply throw
+aways files from old to new when one of the following conditions meets:
+
+- If TTL is set for the FIFO Compaction, RocksDB will throw away files which is
+  older than the configured TTL.
+
+- In case the total DB size is still greater than the configured size, RocksDB
+  will throw away files from the oldest to the newest until the total DB size
+  is under the configured size.
+
+Since FIFO Compaction will throw away files without even taking a look, both
+the TTL and the total DB size must be configured with extra buffer (for
+instance, reserve extra 25% in addition to the target TTL / DB size.)
+
+### Fully Utilize Storage by Properly Rate Limiting
+As mentioned previously, FIFO Compaction deletes the oldest SST file when 1)
+its file creation time is than the specified TTL, or 2) when the total DB size
+exceeds the configured size.  While the TTL condition is what we hope to
+happen constantly, we should avoid trigging the size condition whenever
+possible as it the total amount of data within TTL exceeds the storage size.
+
+To avoid data loss happens due to the size condition, we should set rate
+limiting on writes by `storage size / TTL`.  This will ensure Solana's
+BlockStore handles the write traffic just as fast as how it can digest
+without data loss.
```

Comment:
For context we had some discussion about fifo compaction here in the past with @ryoqun https://discord.com/channels/428295358100013066/439194979856809985/804218651498643457

I think the primary concern being that the FIFO based column would become out of sync with other columns that are also keyed by slot `X` (inconsistent if one column family contains slot `X` and another doesn't). This is avoided by the current compaction filter which removes from all column families based on the same cutoff at slot `X`

Reply:
> Before making any compaction changes, I would be interested in benchmarking RocksDb write throughput vs @steviez's simpler file store.
> 
> If I recall correctly, even without compaction, the writes to Rocks were a fair bit slower (writing directly to the memtables). @steviez, @yhchiang-sol and I have already discussed a few ways to approach this (local benchmarks + cluster benchmarks)

Can I know how was the benchmark conducted?  RocksDB by default will run some compactions otherwise the DB will eventually be full.  If we want to benchmark RocksDB w/o compaction, then we need to config it with FIFO.  Or do I miss anything?

Reply:
FIFO compaction's TTL actually offers the same level of consistency across multiple column families as the current level compaction approach we have now. It is because compactions across different column families are not synchronized. As a result, it is possible that some entries with one slot ID have been dropped by the compaction filter while entries with the same slot ID in other column families are still in the database.